internal/core/swrenderer/draw_functions.rs

```diff
@@ -231,81 +235,105 @@ pub(super) fn draw_rounded_rectangle_line(
-        ((b_ * color1.red() as u32 + a_ * color2.red() as u32) / div) as u8,
-        ((b_ * color1.green() as u32 + a_ * color2.green() as u32) / div) as u8,
-        ((b_ * color1.blue() as u32 + a_ * color2.blue() as u32) / div) as u8,
-    ))
+    PremultipliedRgbaColor {
+        alpha: (m / 255) as u8,
+        red: ((b * color1.red as u32 + a * color2.red as u32) / 255) as u8,
+        green: ((b * color1.green as u32 + a * color2.green as u32) / 255) as u8,
+        blue: ((b * color1.blue as u32 + a * color2.blue as u32) / 255) as u8,
+    }
+}
+
+/// A color whose component have been pre-multiplied by alpha
+///
+/// The renderer operates faster on pre-multiplied color since it
+/// caches the multiplication of its component
+///
+/// PremultipliedRgbaColor can be constructed from a [`Color`] with
+/// the [`From`] trait. This conversion will pre-multiply the color
+/// components
+#[derive(Clone, Copy, Debug, Default)]
+pub struct PremultipliedRgbaColor {
+    pub alpha: u8,
+    pub red: u8,
+    pub green: u8,
+    pub blue: u8,
 }
 
-/// Wrap a color whose component have been pre-multiplied by alpha
-#[derive(Clone, Copy)]
-pub struct Premultiplied(pub Color);
+/// Convert a non-premultiplied color to a premultiplied one
+impl From<Color> for PremultipliedRgbaColor {
+    fn from(col: Color) -> Self {
+        Self::premultiply(col)
+    }
+}
 
-impl Premultiplied {
+impl PremultipliedRgbaColor {
     /// Convert a non premultiplied color to a premultiplied one
-    pub fn premultiply(col: Color) -> Self {
+    fn premultiply(col: Color) -> Self {
         let a = col.alpha() as u16;
-        Self(Color::from_argb_u8(
-            col.alpha(),
-            (col.red() as u16 * a / 255) as u8,
-            (col.green() as u16 * a / 255) as u8,
-            (col.blue() as u16 * a / 255) as u8,
-        ))
+        Self {
+            alpha: col.alpha(),
+            red: (col.red() as u16 * a / 255) as u8,
+            green: (col.green() as u16 * a / 255) as u8,
+            blue: (col.blue() as u16 * a / 255) as u8,
+        }
     }
 }
 
 /// Trait for the pixels in the buffer
 pub trait TargetPixel: Sized + Copy {
-    /// blend a single pixel
-    fn blend_pixel(pix: &mut Self, color: Premultiplied);
-    /// Fill (or blend) the color in the buffer
-    fn blend_buffer(to_fill: &mut [Self], color: Color);
-    /// Create a pixel from the red, gree, blue component in the range 0..=255
-    fn from_rgb(r: u8, g: u8, b: u8) -> Self;
-}
-
-#[cfg(feature = "embedded-graphics")]
-impl TargetPixel for embedded_graphics::pixelcolor::Rgb888 {
-    fn blend_buffer(to_fill: &mut [Self], color: Color) {
-        if color.alpha() == u8::MAX {
-            to_fill.fill(Self::new(color.red(), color.green(), color.blue()))
+    /// Blend a single pixel with a color
+    fn blend(&mut self, color: PremultipliedRgbaColor);
+    /// Fill (or blend) a color in the buffer.
+    fn blend_buffer(to_fill: &mut [Self], color: PremultipliedRgbaColor) {
```

Comment:
I had a thought about this one: I think that we are going to use the term “buffer” (double buffer etc) likely mostly in the context of a two dimensional space. Here it’s closer to a span. Perhaps this should be called `blend_span` or even more Rust style `blend_slice`. What do you think?

Reply:
I like `blend_slice`